### PR TITLE
Adding ability for a job to throw a Non-Retryable exception type

### PR DESCRIPTION
**Pull request by parekhkb**
- Added a NonRetryableException type
- Added support in AutomaticRetryAttribute.cs to prevent retries if a failed job has a exception of type NonRetryableException

### CODE REVIEW

**Comment by odinserj** [comment]
Could you add some details regarding the problem you are facing with and why it can't be solved using the existing functionality?

File: src/Hangfire.Core/Hangfire.Core.csproj [hunk 1]
```diff
@@ -93,6 +93,7 @@
     <Compile Include="Dashboard\Owin\OwinDashboardContextExtensions.cs" />
     <Compile Include="Dashboard\Owin\OwinDashboardRequest.cs" />
     <Compile Include="Dashboard\Owin\OwinDashboardResponse.cs" />
+    <Compile Include="NonRetryableException.cs" />
```
Comment (pieceofsummer) [inline src/Hangfire.Core/Hangfire.Core.csproj]: Also needs to be included in Hangfire.Core.NetStandard.csproj

**Comment by pieceofsummer** [comment]
@odinserj this actually makes sense, since some exceptions shouldn't be retried. For example, the job argument was invalid, or there's some configuration error. Further retries won't make it valid, so there's no reason to keep the job failing again and again.

**Comment by odinserj** [comment]
I'm just willing to add some more details to the PR 😊 Also, there's another implementation of this feature without introducing the new classes with the following API (don't consider property name too seriously):

```csharp
[AutomaticRetry(BlackList = new [] { typeof(InvalidOperationException) })]
```

**Comment by parekhkb** [comment]
We have some cases where some exceptions should not be retried. An example would be that a duplicate primary key exception is encountered when trying to commit to a DB. Such an exception will never succeed no matter how many times you retry. 

I will explore adding a black-list parameter to the constructor.

**Comment by odinserj** [comment]
Thanks for giving the details. I've just proposed the changes, and you will be the first user of this feature, so if you have some argument against it – just let us know!

**Comment by parekhkb** [comment]
@odinserj Thanks! Do you need an Issue to track this? Or do you plan to take in this PR?

File: src/Hangfire.Core/AutomaticRetryAttribute.cs [hunk 2]
```diff
@@ -101,6 +108,26 @@ public AutomaticRetryAttribute()
             Order = 20;
         }
 
+        /// <summary>
+        /// Gets or sets a list of Non-retryable exception types.
+        /// </summary>
+        /// <value>Any non-negative number.</value>
+        /// <exception cref="ArgumentOutOfRangeException">The value in a set operation is less than zero.</exception>
+
+        public Type[] NonRetryable
```
Comment (odinserj) [inline src/Hangfire.Core/AutomaticRetryAttribute.cs]: Looks like there's another word for this – https://en.wiktionary.org/wiki/retriable 😊

**Comment by parekhkb** [comment]
@odinserj, Your earlier comments about proposing the changes. Do you track those somewhere privately? I'm just looking for a ballpark of when to expect the changes. 

Also, I've hit a snag on this PR. the context is never a FailedState, even if I expliclty throw an exception in my job. This is the issue and it might be blocking these changes: https://github.com/HangfireIO/Hangfire/issues/1071

**Pull request closed without merging**